### PR TITLE
PP-586: On a Cray X-series, create a vnode per compute node.

### PR DESCRIPTION
**Pull request by djohnpbs**
#### Issue-ID

* **[PP-586](https://pbspro.atlassian.net/browse/PP-586)**

#### Problem description
* Sites have requested the ability to change the way PBS reports Cray nodes, so that they can have fewer vnodes to deal with instead of multiple vnodes per host.

#### Cause / Analysis
* On a Cray X-series, PBS creates one vnode for each NUMA node reported by the ALPS inventory. Hence sites have to deal with a large number of vnodes.

#### Solution description
* The behavior of PBS will be changed to create, by default, a vnode per compute node as reported in the ALPS inventory. A 'vnode_per_numa_node' variable will provide PBS the ability to choose between creation of a vnode per compute node or a vnode per segment.

#### Checklist:

- [X] I have joined the **[pbspro community forum](http://community.pbspro.org/)**.
- [X] My pull request contains a **single, signed** commit. See **[setting up gpg signature](https://pbspro.atlassian.net/wiki/display/DG/Signing+Your+Git+Commits)**.
- [X] My code follows the **[coding style](https://pbspro.atlassian.net/wiki/display/DG/Coding+Standards)** of this project.
- [X] My change requires project documentation. See **[required documentation checklist](https://pbspro.atlassian.net/wiki/display/DG/Checklist+for+Developing+Features+and+Bug+Fixes)** for details.
- [X] I have added documentation in the **[project documentation area](https://pbspro.atlassian.net/wiki/display/PD)**.
- [ ] I have added new **PTL test(s) to my commit**. (See **[using PTL for testing](https://pbspro.atlassian.net/wiki/display/DG/Using+PTL+for+Testing)**) *(or)*
- [ ] I have added  **manual test(s) to the Jira ticket and explained why PTL is not appropriate** for this case.
- [X] All new and existing automated tests have passed. (See **[running automated PTL tests](https://pbspro.atlassian.net/wiki/display/DG/PTL+Quick+Start+Guide)**).
- [X] I have attached **test logs to the Jira ticket** as evidence of testing/verification.

__***For further information please visit the [Developer Guide Home](https://pbspro.atlassian.net/wiki/display/DG/Developer+Guide+Home).***__

### CODE REVIEW

File: src/resmom/mom_main.c [hunk 1]
```diff
@@ -504,6 +506,7 @@ static struct	specials {
 	{ "alien_kill",			set_alien_kill },
 #if	MOM_ALPS
 	{ "alps_client",		set_alps_client },
+	{ "vnode_per_numa_node", 	set_vnode_per_numa_node },
```
Comment (lisa-altair) [inline src/resmom/mom_main.c]: Nit: If you have to touch the file again, you could make the order here match the order at lines 445-448.

**Comment by lisa-altair** [comment]
Other than my nit, I am fine with the changes.

**Comment by lisa-altair** [comment]
@PBSPro/pbspro-maintainers we know we can't check-in without PTL tests, and those are coming.  In the meantime, we would like to get review feedback on the code so we can address any comments.  If there are none, that's great, and we will focus on the PTL tests.

File: src/resmom/linux/alps.c [hunk 2]
```diff
@@ -3187,12 +3189,28 @@ inventory_to_vnodes(basil_response_t *brp)
...
+
+		/* Only do this for nodes that have accelerators. */
+                if (node->accelerators) {
+                        for (accel=node->accelerators, totaccel=0;
+                                accel; accel=accel->next) {
+                                if (accel->state == basil_accel_state_up)
+                                        /* Only count them if the state is UP */
+                                        totaccel++;
+                        }
+                }
```
Comment (mike0042) [inline src/resmom/linux/alps.c]: Why was this block moved from its original location? The value of totaccel is not used until the next block that tests for node->accelerators

File: src/resmom/linux/alps.c [hunk 3]
```diff
@@ -4208,13 +4277,20 @@ alps_create_reserve_request(job *pjob, basil_request_reserve_t **req)
 
 		/*
 		 * Only match vnodes that begin with mpphost and have
-		 * a following "_<num>_<num>".
+		 * a following "_<num>_<num>" (when vnode_per_numa_node
+		 * is true) otherwise, just plain "_<num>".
```
Comment (mike0042) [inline src/resmom/linux/alps.c]: This comment should be reworded for the sake of clarity.

File: src/resmom/mom_main.c (hunk 1 above)
Comment (mike0042) [inline src/resmom/mom_main.c]: Agreed.

File: src/resmom/mom_main.c [hunk 4]
```diff
@@ -3830,6 +3833,31 @@ set_alps_client(char *value)
...
+ * PBS will create one vnode for the compute node (i.e. it will
+ * cover all segment information in one vnode).
+ *
+ * @retval 0 failure
+ * @retval 1 success
+ */
+static handler_ret_t
+set_vnode_per_numa_node(char *value)
+{
+	static char id[] = "set_vnode_per_numa_node";
```
Comment (mike0042) [inline src/resmom/mom_main.c]: We don't use id anymore. Use `__func__` instead.

File: src/resmom/linux/alps.c (hunk 2 above)
Comment (lisa-altair) [inline src/resmom/linux/alps.c]: By moving it outside of the segment loop we can save ourselves a few cycles this way.  In the BASIL 1.2 XML that we parse, the accelerator information is provided per node, not per segment.  So when the code was further down we were unnecessarily calculating the same accelerator information again and again per segment.

File: src/resmom/linux/alps.c (hunk 3 above)
Comment (lisa-altair) [inline src/resmom/linux/alps.c]: I am happy to reword the comment.  Can you please suggest a change?

File: src/resmom/mom_main.c (hunk 1 above)
Comment (lisa-altair) [inline src/resmom/mom_main.c]: That's 2 votes, I will make that change.

File: src/resmom/mom_main.c (hunk 4 above)
Comment (lisa-altair) [inline src/resmom/mom_main.c]: Will do.

File: src/resmom/linux/alps.c (hunk 3 above)
Comment (mike0042) [inline src/resmom/linux/alps.c]: For this comment, I would suggest: Skip over vnodes where the name does not begin with the expected mpphost string.
Then add another comment as indicated below.

File: src/resmom/linux/alps.c [hunk 5]
```diff
@@ -4208,13 +4277,20 @@ alps_create_reserve_request(job *pjob, basil_request_reserve_t **req)
...
 		 */
 		if (strncmp(vp->vn_vname, mpphost, len) != 0)
 			continue;
 		cp = &vp->vn_vname[len];
-		if (sscanf(cp, "_%ld_%d", &nid, &seg) != 2)
-			continue;
+
```
Comment (mike0042) [inline src/resmom/linux/alps.c]: Please add a comment here: The remainder of the vnode name must match"_<num>_<num>" when vnode_per_numa_node is enabled, or "_<num>" when disabled.

**Comment by lisa-altair** [comment]
Thanks for the suggestions @mike0042.  The changes you requested have been implemented in pull request #233.  I believe we can close this pull request.  However, I am not able to close this pull request.  Can someone please close this?

**Comment by mike0042** [comment]
Moved to https://github.com/PBSPro/pbspro/pull/233

**Pull request closed without merging**